src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -7559,15 +7559,42 @@ public override IOperation VisitRelationalPattern(IRelationalPatternOperation op
 
         public override IOperation VisitBinaryPattern(IBinaryPatternOperation operation, int? argument)
         {
-            return new BinaryPatternOperation(
-                operatorKind: operation.OperatorKind,
-                leftPattern: (IPatternOperation)VisitRequired(operation.LeftPattern),
-                rightPattern: (IPatternOperation)VisitRequired(operation.RightPattern),
-                inputType: operation.InputType,
-                narrowedType: operation.NarrowedType,
-                semanticModel: null,
-                syntax: operation.Syntax,
-                isImplicit: IsImplicit(operation));
+            if (operation.LeftPattern is not IBinaryPatternOperation)
+            {
+                return createOperation(this, operation, (IPatternOperation)VisitRequired(operation.LeftPattern));
+            }
+
+            // Use a manual stack to avoid overflowing on deeply-nested binary patterns
+            var stack = ArrayBuilder<IBinaryPatternOperation>.GetInstance();
+            IBinaryPatternOperation? current = operation;
+
+            do
+            {
+                stack.Push(current);
+                current = current.LeftPattern as IBinaryPatternOperation;
+            } while (current != null);
+
+            var result = (IPatternOperation)VisitRequired(stack.Peek().LeftPattern);
+            while (stack.TryPop(out current))
```

Comment:
>while (stack.TryPop(out current)) [](http://example.com/codeflow?start=12&length=33)

It looks like the stack is never empty on entry, therefore consider using a loop with post-condition instead:
```
            do
            {
                current = stack.Pop()
                result = createOperation(this, current, result);
            }
            while (stack.Any());
```

Reply:
Or perhaps:
```
current = stack.Pop();
var result = (IPatternOperation)VisitRequired(current.LeftPattern);
do
{
    result = createOperation(this, current, result);
}
while (stack.TryPop(out current));

```